semgrep-core/src/analyzing/Dataflow_tainting.ml

```diff
@@ -77,6 +78,49 @@ let option_to_varmap = function
   | Some lvar -> VarMap.singleton (str_of_name lvar) ()
 (*e: function [[Dataflow_tainting.option_to_varmap]] *)
 
+(*****************************************************************************)
+(* Tainted *)
+(*****************************************************************************)
+
+let sanitized config instr =
+  match instr.i with
+  | Call (_, {e=Lvalue({base=Var(("sanitize",_),_);_}); _}, [])
+  | _ when config.is_sanitizer instr
+    -> true
+  | ___else___
+    -> false
+
+let rec tainted env config exp =
+  let go_into = function
+    | Lvalue {base=Var var;_}
+      -> VarMap.mem (str_of_name var) env
+    | Lvalue _
+    | Literal _
+    | FixmeExp _
+      -> false
+    | Composite (_, (_, es, _))
+    | Operator (_, es)
+      -> List.exists (tainted env config) es
+    | Record fields
+      -> List.exists (fun (_, e) -> tainted env config e) fields
+    | Cast (_, e)
+      -> tainted env config e
+  in
+  config.is_source_exp exp || go_into exp.e
+
+let tainted_instr env config instr =
```

Comment:
what was wrong with just using rvars_of_instr and looking if one of them is tainted?
Why the explicit matching of instr cases here? rvars_of_instr was wrong?

Reply:
Note that `go_into` is not recursive, but `tainted` is. Every time we call `tainted` on a subexpression we are checking if the subexpression matches any pattern in `pattern-sources`. I will add a comment to the code.

Reply:
In principle the only interesting subexpressions are the l-values, but we need the `eorig` of those lvalues in order to match. We could have some new `xyz_of_instr` function to return them and it would work as well. But note that the current code allows for _any arbitrary expression_ to be a source of taint... you could e.g. say that `1 + 1` is a source. A more interesting example could be `NULL`  as a source... This could perhaps be interesting for advanced users to play around?